src/GpxDocument.cpp

```diff
@@ -779,1881 +504,575 @@ std::shared_ptr<OsmAnd::GpxDocument::Bounds> OsmAnd::GpxDocument::parseBoundsAtt
+    segment->turnAngle = parser.attributes().value(QStringLiteral(""), QStringLiteral("turnAngle")).toString();
+    segment->types = parser.attributes().value(QStringLiteral(""), QStringLiteral("types")).toString();
+    segment->pointTypes = parser.attributes().value(QStringLiteral(""), QStringLiteral("pointTypes")).toString();
+    segment->names = parser.attributes().value(QStringLiteral(""), QStringLiteral("names")).toString();
+    return segment;
+}
+
+std::shared_ptr<OsmAnd::GeoInfoDocument::RouteType> OsmAnd::GpxDocument::parseRouteTypeAttributes(QXmlStreamReader& parser)
+{
+    const auto type = std::make_shared<RouteType>();
+    type->tag = parser.attributes().value(QStringLiteral(""), QStringLiteral("t")).toString();
+    type->value = parser.attributes().value(QStringLiteral(""), QStringLiteral("v")).toString();
+    return type;
+}
+
+QString OsmAnd::GpxDocument::readText(QXmlStreamReader& xmlReader, QString key)
+{
+    QXmlStreamReader::TokenType tok;
+    QString text;
+    while ((tok = xmlReader.readNext()) != QXmlStreamReader::TokenType::EndDocument)
+    {
+        if (tok == QXmlStreamReader::TokenType::EndElement && xmlReader.name().toString() == key)
+        {
+            break;
+        }
+        else if (tok == QXmlStreamReader::TokenType::Characters)
+        {
+            if (text == nullptr)
+                text = xmlReader.text().toString();
+            else
+                text.append(xmlReader.text().toString());
+        }
+    }
+    return text;
+}
+
+QMap<QString, QString> OsmAnd::GpxDocument::readTextMap(QXmlStreamReader& xmlReader, QString key)
 {
-    std::shared_ptr<GpxDocument> document;
-    std::shared_ptr<GpxMetadata> metadata;
-    std::shared_ptr<Author> author;
-    std::shared_ptr<Copyright> copyright;
-    std::shared_ptr<Bounds> bounds;
-    std::shared_ptr<GpxWpt> wpt;
-    std::shared_ptr<GpxWpt> rpt;
-    std::shared_ptr<GpxTrk> trk;
-    std::shared_ptr<GpxRte> rte;
-    std::shared_ptr<GpxLink> link;
-    std::shared_ptr<GpxRtePt> rtept;
-    std::shared_ptr<GpxTrkPt> trkpt;
-    std::shared_ptr<GpxTrkSeg> trkseg;
-    std::shared_ptr<GpxExtensions> extensions;
-    QStack< std::shared_ptr<GpxExtension> > extensionStack;
-
-    enum class Token
+    QXmlStreamReader::TokenType tok;
+    QString text;
+    QMap<QString, QString> result;
+    while ((tok = xmlReader.readNext()) != QXmlStreamReader::TokenType::EndDocument)
     {
-        gpx,
-        metadata,
-        copyright,
-        author,
-        wpt,
-        trk,
-        rte,
-        link,
-        rtept,
-        trkpt,
-        trkseg,
-        extensions,
-    };
-    QStack<Token> tokens;
+        if (tok == QXmlStreamReader::TokenType::EndElement)
+        {
+            QString tag = xmlReader.name().toString();
+            if (text != nullptr && text.trimmed().length() > 0)
+                result.insert(tag, text);
+            if (tag.compare(key))
```

Comment:
This is not how compare works in QString, please read the documentation and fix this